In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

print('Libraries imported.')

Waiting for a Spark session to start...
Spark Initialization Done! ApplicationId = app-20200413081440-0000
KERNEL_ID = 1ea1c486-7e5c-44cb-8689-bcd4d8204196
Libraries imported.


In [ ]:
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

In [39]:
df = pd.read_html(r'https://www.mapsofindia.com/pincode/india/uttar-pradesh/kanpur-nagar/', header = 1)
df = pd.DataFrame(df[0])
df.head()

Location  Pincode          State      District
0  Agriculture College   208002  Uttar Pradesh  Kanpur Nagar
1                Amour   209401  Uttar Pradesh  Kanpur Nagar
2          Anand Nagar   208011  Uttar Pradesh  Kanpur Nagar
3            Anwarganj   208003  Uttar Pradesh  Kanpur Nagar
4            Aranjhami   209401  Uttar Pradesh  Kanpur Nagar

In [40]:
df.shape

(175, 4)

latitude of Barra, Kanpur, Uttar Pradesh, India is 26.427971, and the longitude is 80.296242.

Arya Nagar Pincode is 208002 With 26.51699 Latitude and 80.31155 Longitude

JNVM Pincode is 208024 With 26.4933 Latitude and 80.2847 Longitude

In [2]:

CLIENT_ID = 'FZV1F5DFXYG3HQX51UIYYUGFZUZASDZ00OM002X5UFT5U0VH' # your Foursquare ID
CLIENT_SECRET = '554WLBIL5FFK3IDA13QH15OMCERX5KZRBZG1M5FKRXH2X5E4' # your Foursquare Secret
VERSION = '20180604'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FZV1F5DFXYG3HQX51UIYYUGFZUZASDZ00OM002X5UFT5U0VH
CLIENT_SECRET:554WLBIL5FFK3IDA13QH15OMCERX5KZRBZG1M5FKRXH2X5E4


In [23]:
radius=5000
LIMIT =100
lat= 26.4933
lng= 80.2847  
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
result = requests.get(url).json()
result

{'meta': {'code': 200, 'requestId': '5e942a46882fc76a3346ab8a'},
 'response': {'headerLocation': 'Kānpur',
  'headerFullLocation': 'Kānpur',
  'headerLocationGranularity': 'city',
  'totalResults': 13,
  'suggestedBounds': {'ne': {'lat': 26.538300045000046,
    'lng': 80.33488625770737},
   'sw': {'lat': 26.448299954999957, 'lng': 80.23451374229263}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4d9cb59ec540236a4e187f4b',
       'name': 'Upper Crust',
       'location': {'address': '112/368 F',
        'crossStreet': 'Swarup Nagar',
        'lat': 26.48222458226252,
        'lng': 80.32085672413587,
        'labeledLatLngs': [{'label': 'display',
          'lat': 26.48222458226252,
          'lng': 80.32085672413587}],
        'distance': 3807,
        'postalCode': '

In [ ]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [33]:
venues = result['response']['groups'][0]['items']
    
venues_df = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng', 'venue.location.distance']
venues_df = venues_df.loc[:, filtered_columns]

# filter the category for each row
venues_df['venue.categories'] = venues_df.apply(get_category_type, axis=1)

venues_df.head()

venue.name      venue.categories  venue.location.lat  \
0         Upper Crust                Bakery           26.482225   
1   Hotel Royal Cliff                 Hotel           26.479289   
2  Aromas, Arya Nagar  Fast Food Restaurant           26.482953   
3         Rave @ Moti         Shopping Mall           26.481579   
4              Rave 3         Shopping Mall           26.492315   

   venue.location.lng  venue.location.distance  
0           80.320857                     3807  
1           80.314855                     3385  
2           80.318024                     3514  
3           80.300102                     2014  
4           80.327160                     4231

In [34]:
venues_df = venues_df.rename(columns = {'venue.name':'Venue','venue.categories':'Category','venue.location.lat':'Latitude','venue.location.lng':'Longitude', 'venue.location.distance': 'Distance'})
venues_df.head()

Venue              Category   Latitude  Longitude  Distance
0         Upper Crust                Bakery  26.482225  80.320857      3807
1   Hotel Royal Cliff                 Hotel  26.479289  80.314855      3385
2  Aromas, Arya Nagar  Fast Food Restaurant  26.482953  80.318024      3514
3         Rave @ Moti         Shopping Mall  26.481579  80.300102      2014
4              Rave 3         Shopping Mall  26.492315  80.327160      4231

In [2]:
#Reading in the table and setting the header to the first row
df = pd.read_html(r'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header = 0)
df = pd.DataFrame(df[0])
df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

In [3]:
cdf= df[df['Borough']!='Not assigned']
cdf.reset_index(inplace = True, drop = True)

In [4]:
cdf_fil= cdf[cdf['Borough'].str.contains('Toronto')]
cdf_fil.head(20)

Postal code           Borough  \
2          M5A  Downtown Toronto   
4          M7A  Downtown Toronto   
9          M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
54         M4M      East Toronto   
61         M4N   Central Toronto   
62         M5N   Central Toronto   

                                         Neighborhood  
2                          Regent Park / Harbourfront  
4        Queen's Park / Ontario Provincial Government  
9                            Garden District, Ryerson  
15                                     St. James Town  
19                                        The Beaches  
20                                        Berczy Park  
24                                 Central Bay Street  
25                                           Christie  
30                         Richmond / Adelaide / King  
31                      Dufferin / Dovercourt Village  
36  Harbourfront East / Union Station / Toronto Is...  
37                          Little Portugal / Trinity  
41                      The Danforth West / Riverdale  
42          Toronto Dominion Centre / Design Exchange  
43     Brockton / Parkdale Village / Exhibition Place  
47                    India Bazaar / The Beaches West  
48                    Commerce Court / Victoria Hotel  
54                                    Studio District  
61                                      Lawrence Park  
62                                           Roselawn

In [5]:
#Replacing the "/" with a ","
cdf_fil['Neighborhood'] = cdf_fil['Neighborhood'].replace("/", ",", regex = True)
cdf_fil.head()

/opt/ibm/conda/miniconda3.6/lib/python3.6/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  from ipykernel import kernelapp as app


Postal code           Borough                                  Neighborhood
2          M5A  Downtown Toronto                    Regent Park , Harbourfront
4          M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government
9          M5B  Downtown Toronto                      Garden District, Ryerson
15         M5C  Downtown Toronto                                St. James Town
19         M4E      East Toronto                                   The Beaches

In [6]:
for i in cdf_fil['Neighborhood']:
    if i=='NaN':
      print (i)

In [7]:
cdf_fil.shape

(39, 3)

In [8]:
## Importing the CSV file with geo-loc data
geo = pd.read_csv(r'http://cocl.us/Geospatial_data', header = 0)
geo.shape

(103, 3)

In [9]:
geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [10]:
pdf= pd.DataFrame(columns= ['Postal Code', 'Borough', 'Neighborhood', 'Latitude', 'Longitude'])

In [11]:
for i in cdf['Postal code']:
    for j in geo['Postal Code']:
        if i==j:
               pdf['Postal Code']=cdf['Postal code']
               pdf['Borough']=cdf['Borough']
               pdf['Neighborhood']=cdf['Neighborhood']
               pdf['Latitude']= geo['Latitude']
               pdf['Longitude']=geo['Longitude']

In [12]:
pdf= pdf[pdf['Borough'].str.contains('Toronto')]
pdf.head()

Postal Code           Borough  \
2          M5A  Downtown Toronto   
4          M7A  Downtown Toronto   
9          M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
19         M4E      East Toronto   

                                    Neighborhood   Latitude  Longitude  
2                     Regent Park / Harbourfront  43.763573 -79.188711  
4   Queen's Park / Ontario Provincial Government  43.773136 -79.239476  
9                       Garden District, Ryerson  43.692657 -79.264848  
15                                St. James Town  43.799525 -79.318389  
19                                   The Beaches  43.786947 -79.385975

In [13]:
len(pdf['Neighborhood'].unique())

39

In [14]:
pdf['Borough'].unique()

array(['Downtown Toronto', 'East Toronto', 'West Toronto',
       'Central Toronto'], dtype=object)

In [15]:
pdf.shape

(39, 5)

In [16]:
CLIENT_ID = 'FZV1F5DFXYG3HQX51UIYYUGFZUZASDZ00OM002X5UFT5U0VH' # your Foursquare ID
CLIENT_SECRET = '554WLBIL5FFK3IDA13QH15OMCERX5KZRBZG1M5FKRXH2X5E4' # your Foursquare Secret
VERSION = '20180604'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FZV1F5DFXYG3HQX51UIYYUGFZUZASDZ00OM002X5UFT5U0VH
CLIENT_SECRET:554WLBIL5FFK3IDA13QH15OMCERX5KZRBZG1M5FKRXH2X5E4


In [17]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [18]:

venues_list=[]
radius=500
LIMIT =100

for name, lat, lng in zip(pdf['Neighborhood'], pdf['Latitude'], pdf['Longitude']):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            x['venue']['name'], 
            x['venue']['location']['lat'], 
            x['venue']['location']['lng'],  
            x['venue']['categories'][0]['name']) for x in results])

ven_pdf = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
ven_pdf.columns = ['Neighborhood', 
                  'Latitude', 
                  'Longitude', 
                  'venue', 
                  'Venue latitude', 
                  'Venue longitude', 
                  'categories']

ven_pdf.head(20)

Neighborhood   Latitude  Longitude  \
0                     Regent Park / Harbourfront  43.763573 -79.188711   
1                     Regent Park / Harbourfront  43.763573 -79.188711   
2                     Regent Park / Harbourfront  43.763573 -79.188711   
3                     Regent Park / Harbourfront  43.763573 -79.188711   
4                     Regent Park / Harbourfront  43.763573 -79.188711   
5                     Regent Park / Harbourfront  43.763573 -79.188711   
6                     Regent Park / Harbourfront  43.763573 -79.188711   
7   Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
8   Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
9   Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
10  Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
11  Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
12  Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
13  Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
14  Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
15  Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
16                      Garden District, Ryerson  43.692657 -79.264848   
17                      Garden District, Ryerson  43.692657 -79.264848   
18                      Garden District, Ryerson  43.692657 -79.264848   
19                      Garden District, Ryerson  43.692657 -79.264848   

                           venue  Venue latitude  Venue longitude  \
0              G & G Electronics       43.765309       -79.191537   
1               Big Bite Burrito       43.766299       -79.190720   
2          Enterprise Rent-A-Car       43.764076       -79.193406   
3                 RBC Royal Bank       43.766790       -79.191151   
4          Woburn Medical Centre       43.766631       -79.192286   
5   Lawrence Ave E & Kingston Rd       43.767704       -79.189490   
6                       Eggsmart       43.767800       -79.190466   
7            Federick Restaurant       43.774697       -79.241142   
8       Drupati's Roti & Doubles       43.775222       -79.241678   
9                    Thai One On       43.774468       -79.241268   
10  Centennial Recreation Centre       43.774593       -79.236500   
11               TD Canada Trust       43.774830       -79.241251   
12                  Petro-Canada       43.774106       -79.243097   
13                    B&A Bakery       43.774391       -79.243877   
14                    The Avenue       43.772437       -79.241962   
15     Popeyes Louisiana Kitchen       43.775930       -79.235328   
16                The Birchcliff       43.691666       -79.264532   
17   Birchmount Community Centre       43.695175       -79.262161   
18           Scarborough Gardens       43.694647       -79.262230   
19            Birchmount Stadium       43.695323       -79.261293   

               categories  
0       Electronics Store  
1      Mexican Restaurant  
2     Rental Car Location  
3                    Bank  
4          Medical Center  
5            Intersection  
6          Breakfast Spot  
7        Hakka Restaurant  
8    Caribbean Restaurant  
9         Thai Restaurant  
10     Athletics & Sports  
11                   Bank  
12            Gas Station  
13                 Bakery  
14                 Lounge  
15    Fried Chicken Joint  
16                   Café  
17  General Entertainment  
18           Skating Rink  
19        College Stadium

In [19]:
ven_pdf.groupby('Neighborhood').count()

Latitude  Longitude  \
Neighborhood                                                              
Berczy Park                                                1          1   
Brockton / Parkdale Village / Exhibition Place            40         40   
Business reply mail Processing CentrE                      4          4   
CN Tower / King and Spadina / Railway Lands / H...        18         18   
Central Bay Street                                         8          8   
Christie                                                   2          2   
Church and Wellesley                                       7          7   
Commerce Court / Victoria Hotel                            1          1   
Davisville                                                 4          4   
Davisville North                                          66         66   
Dufferin / Dovercourt Village                              4          4   
First Canadian Place / Underground city                    2          2   
Forest Hill North & West                                  16         16   
Garden District, Ryerson                                   4          4   
Harbourfront East / Union Station / Toronto Isl...        10         10   
High Park / The Junction South                            97         97   
India Bazaar / The Beaches West                           31         31   
Kensington Market / Chinatown / Grange Park               37         37   
Lawrence Park                                            100        100   
Little Portugal / Trinity                                  4          4   
Moore Park / Summerhill East                              14         14   
North Toronto West                                         4          4   
Parkdale / Roncesvalles                                   18         18   
Queen's Park / Ontario Provincial Government               9          9   
Regent Park / Harbourfront                                 7          7   
Richmond / Adelaide / King                                 3          3   
Rosedale                                                   2          2   
Roselawn                                                  28         28   
Runnymede / Swansea                                        5          5   
St. James Town                                            14         14   
St. James Town / Cabbagetown                               2          2   
Stn A PO Boxes                                            14         14   
Studio District                                          100        100   
Summerhill West / Rathnelly / South Hill / Fore...        13         13   
The Annex / North Midtown / Yorkville                      4          4   
The Beaches                                                4          4   
The Danforth West / Riverdale                             42         42   
Toronto Dominion Centre / Design Exchange                 22         22   
University of Toronto / Harbord                            4          4   

                                                    venue  Venue latitude  \
Neighborhood                                                                
Berczy Park                                             1               1   
Brockton / Parkdale Village / Exhibition Place         40              40   
Business reply mail Processing CentrE                   4               4   
CN Tower / King and Spadina / Railway Lands / H...     18              18   
Central Bay Street                                      8               8   
Christie                                                2               2   
Church and Wellesley                                    7               7   
Commerce Court / Victoria Hotel                         1               1   
Davisville                                              4               4   
Davisville North                                       66              66   
Dufferin / Dovercourt Village                       

In [20]:
# one hot encoding
onehot = pd.get_dummies(ven_pdf[['categories']], prefix="", prefix_sep="")

In [21]:
# add neighborhood column back to dataframe
onehot['Neighborhood'] = ven_pdf['Neighborhood']

In [22]:
onehot.head()

Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Art Gallery  \
0                 0                    0             0            0   
1                 0                    0             0            0   
2                 0                    0             0            0   
3                 0                    0             0            0   
4                 0                    0             0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     1    0               0   
4          0           0           0       0     0    0               0   

   Basketball Court  Basketball Stadium  Bed & Breakfast  Beer Bar  \
0                 0                   0                0         0   
1                 0                   0                0         0   
2                 0                   0                0         0   
3                 0                   0                0         0   
4                 0                   0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0           0                   0       0           0              0   
1           0                   0       0           0              0   
2           0                   0       0           0              0   
3           0                   0       0           0              0   
4           0                   0       0           0              0   

   Bookstore  Boutique  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0          0         0               0        0                0         0   
1          0         0               0        0                0         0   
2          0         0               0        0                0         0   
3          0         0               0        0                0         0   
4          0         0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Stop  Butcher  Cafeteria  Café  \
0             0              0         0         0        0          0     0   
1             0              0         0         0        0          0     0   
2             0              0         0         0        0          0     0   
3             0              0         0         0        0          0     0   
4             0              0         0         0        0          0     0   

   Camera Store  Candy Store  Caribbean Restaurant  Cheese Shop  \
0             0            0                     0            0   
1             0            0                     0            0   
2             0            0                     0            0   
3             0            0                     0            0   
4             0         

In [23]:
# move neighborhood column to the first column
onehot.set_index('Neighborhood', inplace = True)

In [24]:
onehot.head(20)

Airport  Airport Food Court  \
Neighborhood                                                                
Regent Park / Harbourfront                          0                   0   
Regent Park / Harbourfront                          0                   0   
Regent Park / Harbourfront                          0                   0   
Regent Park / Harbourfront                          0                   0   
Regent Park / Harbourfront                          0                   0   
Regent Park / Harbourfront                          0                   0   
Regent Park / Harbourfront                          0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Queen's Park / Ontario Provincial Government        0                   0   
Garden District, Ryerson                            0                   0   
Garden District, Ryerson                            0                   0   
Garden District, Ryerson                            0                   0   
Garden District, Ryerson                            0                   0   

                                              Airport Gate  Airport Lounge  \
Neighborhood                                                                 
Regent Park / Harbourfront                               0               0   
Regent Park / Harbourfront                               0               0   
Regent Park / Harbourfront                               0               0   
Regent Park / Harbourfront                               0               0   
Regent Park / Harbourfront                               0               0   
Regent Park / Harbourfront                               0               0   
Regent Park / Harbourfront                               0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Queen's Park / Ontario Provincial Government             0               0   
Garden District, Ryerson                                 0               0   
Garden District, Ryerson                                 0               0   
Garden District, Ryerson                                 0               0   
Garden District, Ryerson                                 0               0   

                                              Airport Service  \
Neighborhood                                                    
Regent Park / Harbourfront                                  0   
Regent Park / Harbourfront                                  0   
Regent Park / Harbourfront                                  0   
Regent Park / Harbourfront                                  0   
Regent Park / Harbourfront                                  0   
Regent Park / Harbourfront                                  0   
Regent Park / Harbourfront                                  0   
Queen's Park / Ontario Provincial Government     

In [25]:
grouped = onehot.groupby('Neighborhood').mean().reset_index()
grouped

Neighborhood   Airport  \
0                                         Berczy Park  0.000000   
1      Brockton / Parkdale Village / Exhibition Place  0.000000   
2               Business reply mail Processing CentrE  0.000000   
3   CN Tower / King and Spadina / Railway Lands / ...  0.000000   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                     Commerce Court / Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                      Dufferin / Dovercourt Village  0.000000   
11            First Canadian Place / Underground city  0.000000   
12                           Forest Hill North & West  0.062500   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East / Union Station / Toronto Is...  0.000000   
15                     High Park / The Junction South  0.000000   
16                    India Bazaar / The Beaches West  0.000000   
17        Kensington Market / Chinatown / Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                          Little Portugal / Trinity  0.000000   
20                       Moore Park / Summerhill East  0.000000   
21                                 North Toronto West  0.000000   
22                            Parkdale / Roncesvalles  0.000000   
23       Queen's Park / Ontario Provincial Government  0.000000   
24                         Regent Park / Harbourfront  0.000000   
25                         Richmond / Adelaide / King  0.333333   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                Runnymede / Swansea  0.000000   
29                                     St. James Town  0.000000   
30                       St. James Town / Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West / Rathnelly / South Hill / For...  0.000000   
34              The Annex / North Midtown / Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                      The Danforth West / Riverdale  0.000000   
37          Toronto Dominion Centre / Design Exchange  0.000000   
38                    University of Toronto / Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0               0.0000        0.0000           0.000            0.000   
1               0.0000        0.0000           0.000            0.000   
2               0.0000        0.0000           0.000            0.000   
3               0.0000        0.0000           0.000            0.000   
4               0.0000        0.0000           0.000            0.000   
5               0.0000        0.0000           0.000            0.000   
6               0.0000        0.0000           0.000            0.000   
7               0.0000        0.0000           0.000            0.000   
8               0.0000        0.0000           0.000            0.000   
9               0.0000        0.0000           0.000            0.000   
10              0.0000        0.0000           0.000            0.000   
11              0.0000        0.0000           0.000            0.000   
12              0.0625        0.0625           0.125            0.125   
13              0.0000        0.0000           0.000            0.000   
14              0.0000        0.0000           0.000            0.000   
15              0.0000        0.0000           0.000            0.000   
16              0.0000        0.0000           0.000            0.000   
17              0.0000        0.0000          

In [26]:
grouped.drop('Neighborhood', axis=1, inplace=True)
grouped.head()

Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Art Gallery  \
0               0.0                 0.00           0.0          0.0   
1               0.0                 0.05           0.0          0.0   
2               0.0                 0.00           0.0          0.0   
3               0.0                 0.00           0.0          0.0   
4               0.0                 0.00           0.0          0.0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                  0.0               0.0                 0.0       0.000000   
1                  0.0               0.0                 0.0       0.000000   
2                  0.0               0.0                 0.0       0.000000   
3                  0.0               0.0                 0.0       0.055556   
4                  0.0               0.0                 0.0       0.000000   

   BBQ Joint  Baby Store  Bagel Shop  Bakery   Bank    Bar  Baseball Field  \
0        0.0         0.0         0.0    0.00  0.000  0.000             0.0   
1        0.0         0.0         0.0    0.05  0.025  0.025             0.0   
2        0.0         0.0         0.0    0.00  0.000  0.000             0.0   
3        0.0         0.0         0.0    0.00  0.000  0.000             0.0   
4        0.0         0.0         0.0    0.00  0.125  0.000             0.0   

   Basketball Court  Basketball Stadium  Bed & Breakfast  Beer Bar  \
0               0.0                 0.0              0.0       0.0   
1               0.0                 0.0              0.0       0.0   
2               0.0                 0.0              0.0       0.0   
3               0.0                 0.0              0.0       0.0   
4               0.0                 0.0              0.0       0.0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0         0.0                 0.0     0.0         0.0            0.0   
1         0.0                 0.0     0.0         0.0            0.0   
2         0.0                 0.0     0.0         0.0            0.0   
3         0.0                 0.0     0.0         0.0            0.0   
4         0.0                 0.0     0.0         0.0            0.0   

   Bookstore  Boutique  Breakfast Spot   Brewery  Bubble Tea Shop  Building  \
0      0.000       0.0             0.0  0.000000              0.0       0.0   
1      0.025       0.0             0.0  0.050000              0.0       0.0   
2      0.000       0.0             0.0  0.000000              0.0       0.0   
3      0.000       0.0             0.0  0.055556              0.0       0.0   
4      0.000       0.0             0.0  0.000000              0.0       0.0   

   Burger Joint  Burrito Place  Bus Line  Bus Stop  Butcher  Cafeteria  Café  \
0           0.0       0.000000      0.00       0.0    0.000        1.0   0.0   
1           0.0       0.000000      0.00       0.0    0.000        0.0   0.1   
2           0.0       0.000000      0.25       0.0    0.000        0.0   0.0   
3           0.0       0.055556      0.00       0.0    0.000        0.0   0.0   
4           0.0       0.000000      0.00       0.0    0.125        0.0   0.0   

   Camera Store  Candy Store  Caribbean Restaurant  Cheese Shop  \
0           0.0          0.0                   0.0        0.000   
1           0.0          0.0                   0.0        0.025   
2           0.0          0.0                   0.0        0.000   
3           0.0          0.0                   0.0        0.000   


In [27]:
# set number of clusters
clusters = 5

# run k-means clustering
kmeans = KMeans(n_clusters=clusters, random_state=0).fit(grouped)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 0, 0, 0, 0, 2, 0, 1, 0, 0], dtype=int32)

In [28]:
grouped.shape
ven_pdf.shape
pdf.shape
len(kmeans.labels_)

39

In [29]:
groupc = onehot.groupby('Neighborhood').mean().reset_index()
groupc.head()

Neighborhood  Airport  \
0                                        Berczy Park      0.0   
1     Brockton / Parkdale Village / Exhibition Place      0.0   
2              Business reply mail Processing CentrE      0.0   
3  CN Tower / King and Spadina / Railway Lands / ...      0.0   
4                                 Central Bay Street      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Art Gallery  \
0               0.0                 0.00           0.0          0.0   
1               0.0                 0.05           0.0          0.0   
2               0.0                 0.00           0.0          0.0   
3               0.0                 0.00           0.0          0.0   
4               0.0                 0.00           0.0          0.0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                  0.0               0.0                 0.0       0.000000   
1                  0.0               0.0                 0.0       0.000000   
2                  0.0               0.0                 0.0       0.000000   
3                  0.0               0.0                 0.0       0.055556   
4                  0.0               0.0                 0.0       0.000000   

   BBQ Joint  Baby Store  Bagel Shop  Bakery   Bank    Bar  Baseball Field  \
0        0.0         0.0         0.0    0.00  0.000  0.000             0.0   
1        0.0         0.0         0.0    0.05  0.025  0.025             0.0   
2        0.0         0.0         0.0    0.00  0.000  0.000             0.0   
3        0.0         0.0         0.0    0.00  0.000  0.000             0.0   
4        0.0         0.0         0.0    0.00  0.125  0.000             0.0   

   Basketball Court  Basketball Stadium  Bed & Breakfast  Beer Bar  \
0               0.0                 0.0              0.0       0.0   
1               0.0                 0.0              0.0       0.0   
2               0.0                 0.0              0.0       0.0   
3               0.0                 0.0              0.0       0.0   
4               0.0                 0.0              0.0       0.0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0         0.0                 0.0     0.0         0.0            0.0   
1         0.0                 0.0     0.0         0.0            0.0   
2         0.0                 0.0     0.0         0.0            0.0   
3         0.0                 0.0     0.0         0.0            0.0   
4         0.0                 0.0     0.0         0.0            0.0   

   Bookstore  Boutique  Breakfast Spot   Brewery  Bubble Tea Shop  Building  \
0      0.000       0.0             0.0  0.000000              0.0       0.0   
1      0.025       0.0             0.0  0.050000              0.0       0.0   
2      0.000       0.0             0.0  0.000000              0.0       0.0   
3      0.000       0.0             0.0  0.055556              0.0       0.0   
4      0.000       0.0             0.0  0.000000              0.0       0.0   

   Burger Joint  Burrito Place  Bus Line  Bus Stop  Butcher  Cafeteria  Café  \
0           0.0       0.000000      0.00       0.0    0.000        1.0   0.0   
1           0.0       0.000000      0.00       0.0    0.000        0.0   0.1   
2           0.0       0.000000      0.25       0.0    0.000        0.0   0.0   
3           0.0       0.055556      0.00       0.0    0.000        0.0   0.0   
4           0.0       0.000000      0.00       0.0    0.125        0.0   0.0   

   Camera Store  Candy Store  Carib

In [30]:
num_top_venues = 5

for c in groupc['Neighborhood']:
    print("----"+c+"----")
    temp = groupc[groupc['Neighborhood'] == c].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
               venue  freq
0          Cafeteria   1.0
1            Airport   0.0
2  Mobile Phone Shop   0.0
3     Lingerie Store   0.0
4       Liquor Store   0.0


----Brockton / Parkdale Village / Exhibition Place----
         venue  freq
0         Café  0.10
1  Coffee Shop  0.08
2    Gastropub  0.05
3      Brewery  0.05
4       Bakery  0.05


----Business reply mail Processing CentrE----
                venue  freq
0   Mobile Phone Shop  0.25
1         Pizza Place  0.25
2      Sandwich Place  0.25
3            Bus Line  0.25
4  Miscellaneous Shop  0.00


----CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst  Quay / South Niagara / Island airport----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2                 Spa  0.06
3       Garden Center  0.06
4              Garden  0.06


----Central Bay Street----
           venue  freq
0           Bank  0.12
1  Grocery Store  0.12
2       Pharmacy  0.12
3    C

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [32]:
num_top_venues = 10

# create columns according to number of top venues
columns = ['Neighborhood']

indicators = ['st', 'nd', 'rd']
for i in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(i+1, indicators[i]))
    except:
        columns.append('{}th Most Common Venue'.format(i+1))


In [33]:
# create a new dataframe
sorted = pd.DataFrame(columns=columns)
sorted['Neighborhood'] = groupc['Neighborhood']

for ind in np.arange(groupc.shape[0]):
    sorted.iloc[ind, 1:] = return_most_common_venues(groupc.iloc[ind, :], num_top_venues)

sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park             Cafeteria   
1     Brockton / Parkdale Village / Exhibition Place                  Café   
2              Business reply mail Processing CentrE              Bus Line   
3  CN Tower / King and Spadina / Railway Lands / ...    Light Rail Station   
4                                 Central Bay Street          Home Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Yoga Studio         Deli / Bodega   Empanada Restaurant   
1           Coffee Shop             Gastropub                Bakery   
2        Sandwich Place           Pizza Place     Mobile Phone Shop   
3           Yoga Studio         Auto Workshop      Recording Studio   
4               Butcher           Coffee Shop         Grocery Store   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
1               Brewery          American Restaurant           Yoga Studio   
2           Yoga Studio  Eastern European Restaurant   Dumpling Restaurant   
3               Brewery                   Skate Park            Smoke Shop   
4        Discount Store                     Pharmacy           Pizza Place   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                      Dog Run         Discount Store  
1        Ice Cream Shop                    Bookstore     Italian Restaurant  
2      Doner Restaurant                      Dog Run         Discount Store  
3                   Spa               Farmers Market   Fast Food Restaurant  
4                  Bank  Eastern European Restaurant    Dumpling Restaurant

In [34]:
# add clustering labels
sorted.insert(0, 'Cluster Labels', kmeans.labels_)
sorted.head()

Cluster Labels                                       Neighborhood  \
0               3                                        Berczy Park   
1               0     Brockton / Parkdale Village / Exhibition Place   
2               0              Business reply mail Processing CentrE   
3               0  CN Tower / King and Spadina / Railway Lands / ...   
4               0                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0             Cafeteria           Yoga Studio         Deli / Bodega   
1                  Café           Coffee Shop             Gastropub   
2              Bus Line        Sandwich Place           Pizza Place   
3    Light Rail Station           Yoga Studio         Auto Workshop   
4          Home Service               Butcher           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue        6th Most Common Venue  \
0   Empanada Restaurant     Electronics Store  Eastern European Restaurant   
1                Bakery               Brewery          American Restaurant   
2     Mobile Phone Shop           Yoga Studio  Eastern European Restaurant   
3      Recording Studio               Brewery                   Skate Park   
4         Grocery Store        Discount Store                     Pharmacy   

  7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
0   Dumpling Restaurant      Doner Restaurant                      Dog Run   
1           Yoga Studio        Ice Cream Shop                    Bookstore   
2   Dumpling Restaurant      Doner Restaurant                      Dog Run   
3            Smoke Shop                   Spa               Farmers Market   
4           Pizza Place                  Bank  Eastern European Restaurant   

  10th Most Common Venue  
0         Discount Store  
1     Italian Restaurant  
2         Discount Store  
3   Fast Food Restaurant  
4    Dumpling Restaurant

In [35]:
sorted.set_index('Neighborhood')

Cluster Labels  \
Neighborhood                                                         
Berczy Park                                                      3   
Brockton / Parkdale Village / Exhibition Place                   0   
Business reply mail Processing CentrE                            0   
CN Tower / King and Spadina / Railway Lands / H...               0   
Central Bay Street                                               0   
Christie                                                         2   
Church and Wellesley                                             0   
Commerce Court / Victoria Hotel                                  1   
Davisville                                                       0   
Davisville North                                                 0   
Dufferin / Dovercourt Village                                    0   
First Canadian Place / Underground city                          4   
Forest Hill North & West                                         0   
Garden District, Ryerson                                         0   
Harbourfront East / Union Station / Toronto Isl...               0   
High Park / The Junction South                                   0   
India Bazaar / The Beaches West                                  0   
Kensington Market / Chinatown / Grange Park                      0   
Lawrence Park                                                    0   
Little Portugal / Trinity                                        0   
Moore Park / Summerhill East                                     0   
North Toronto West                                               2   
Parkdale / Roncesvalles                                          0   
Queen's Park / Ontario Provincial Government                     0   
Regent Park / Harbourfront                                       0   
Richmond / Adelaide / King                                       2   
Rosedale                                                         4   
Roselawn                                                         0   
Runnymede / Swansea                                              0   
St. James Town                                                   0   
St. James Town / Cabbagetown                                     0   
Stn A PO Boxes                                                   0   
Studio District                                                  0   
Summerhill West / Rathnelly / South Hill / Fore...               0   
The Annex / North Midtown / Yorkville                            2   
The Beaches                                                      0   
The Danforth West / Riverdale                                    0   
Toronto Dominion Centre / Design Exchange                        0   
University of Toronto / Harbord                                  0   

                                                   1st Most Common Venue  \
Neighborhood                                                               
Berczy Park                                                    Cafeteria   
Brockton / Parkdale Village / Exhibition Place                      Café   
Business reply mail Processing CentrE                           Bus Line   
CN Tower / King and Spadina / Railway Lands / H...    Light Rail Station   
Central Bay Street                                          Home Service   
Christie                                                            Park   
Church and Wellesley                                         Pizza Place   
Commerce Court / Victoria Hotel                               Restaurant   
Davisville                                                          Park   
Davisville North                                             Coffee Shop   
Dufferin / Dovercourt Village                              Grocery Store   
First Canadian Place / Underground city                      Fabric Shop   
Forest Hill North & West                                  Airport Lounge   
Garden District, Ryerson                 

In [36]:
merged = pdf

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
merged = merged.join(sorted.set_index('Neighborhood'), on='Neighborhood')

merged.head() # check the last columns!

Postal Code           Borough  \
2          M5A  Downtown Toronto   
4          M7A  Downtown Toronto   
9          M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
19         M4E      East Toronto   

                                    Neighborhood   Latitude  Longitude  \
2                     Regent Park / Harbourfront  43.763573 -79.188711   
4   Queen's Park / Ontario Provincial Government  43.773136 -79.239476   
9                       Garden District, Ryerson  43.692657 -79.264848   
15                                St. James Town  43.799525 -79.318389   
19                                   The Beaches  43.786947 -79.385975   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2                0     Electronics Store        Medical Center   
4                0                Lounge      Hakka Restaurant   
9                0          Skating Rink                  Café   
15               0  Fast Food Restaurant    Chinese Restaurant   
19               0   Japanese Restaurant    Chinese Restaurant   

   3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
2         Breakfast Spot           Intersection   Rental Car Location   
4    Fried Chicken Joint     Athletics & Sports       Thai Restaurant   
9        College Stadium  General Entertainment               Gay Bar   
15            Nail Salon               Pharmacy        Breakfast Spot   
19                  Bank                   Café           Yoga Studio   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
2                   Bank    Mexican Restaurant  Eastern European Restaurant   
4   Caribbean Restaurant                  Bank                       Bakery   
9           Dance Studio     Electronics Store  Eastern European Restaurant   
15                  Bank           Supermarket                 Camera Store   
19          Dessert Shop   Empanada Restaurant            Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
2           Dumpling Restaurant       Doner Restaurant  
4                   Gas Station                Dog Run  
9           Dumpling Restaurant       Doner Restaurant  
15                  Pizza Place         Sandwich Place  
19  Eastern European Restaurant    Dumpling Restaurant

In [38]:
# create map
for lat, lon in zip(merged['Latitude'], merged['Longitude']):
    map_clusters = folium.Map(location=[lat, lon], zoom_start=11)

# set color scheme for the clusters
x = np.arange(clusters)
ys = [i + x + (i*x)**2 for i in range(clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
       
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(merged['Latitude'], merged['Longitude'], merged['Neighborhood'], merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
    
map_clusters